### ISSUE TITLE
Every time we remove doctrine subscriber, we can no longer clear the cache

### ISSUE THREAD

**Issue by ostrolucky** [symfony/symfony#22243]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | master

This problem goes as far back as I remember. When we remove or change the name of some of the doctrine subscriber, symfony cannot rebuild container. Apparently container tries to fetch old FCQN even when running cache:clear. This is a big problem for us, because our deployments are not reliable. We cannot replace it with rm -rf var/cache/*, because that one is also not reliable (it has concurrency issues). I have prepared minimal containing example of the issue. Apparently ACL has something to do with it. 

Here's what happens:

```
$ bin/console cache:clear --no-warmup -vvv

                                                           
  [Symfony\Component\Debug\Exception\FatalThrowableError]  
  Class 'SomeSubscriber' not found                         
                                                           

Exception trace:
 () at /home/gadelat/Documents/ATS/app/cache/prod/appProdProjectContainer.php:991
 appProdProjectContainer->getSome_SubscriberService() at /home/gadelat/Documents/ATS/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Container.php:341
 Symfony\Component\DependencyInjection\Container->get() at /home/gadelat/Documents/ATS/app/cache/prod/appProdProjectContainer.php:369
 appProdProjectContainer->getDoctrine_Dbal_DefaultConnectionService() at /home/gadelat/Documents/ATS/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Container.php:341
 Symfony\Component\DependencyInjection\Container->get() at /home/gadelat/Documents/ATS/app/cache/prod/appProdProjectContainer.php:731
 appProdProjectContainer->getSecurity_Acl_ProviderService() at /home/gadelat/Documents/ATS/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Container.php:341
 Symfony\Component\DependencyInjection\Container->get() at /home/gadelat/Documents/ATS/vendor/symfony/symfony/src/Symfony/Bundle/SecurityBundle/Command/SetAclCommand.php:42
 Symfony\Bundle\SecurityBundle\Command\SetAclCommand->isEnabled() at /home/gadelat/Documents/ATS/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:417
 Symfony\Component\Console\Application->add() at /home/gadelat/Documents/ATS/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Bundle/Bundle.php:193
 Symfony\Component\HttpKernel\Bundle\Bundle->registerCommands() at /home/gadelat/Documents/ATS/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:135
 Symfony\Bundle\FrameworkBundle\Console\Application->registerCommands() at /home/gadelat/Documents/ATS/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:108
 Symfony\Bundle\FrameworkBundle\Console\Application->all() at /home/gadelat/Documents/ATS/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:72
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /home/gadelat/Documents/ATS/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:124
 Symfony\Component\Console\Application->run() at /home/gadelat/Documents/ATS/bin/console:6
```
---
Here's minimal reproducible code:

> bin/console
```php
#!/usr/bin/env php
<?php
$loader = require __DIR__.'/../app/autoload.php';

$application = new Symfony\Bundle\FrameworkBundle\Console\Application(new AppKernel('prod', false));
$application->run(new Symfony\Component\Console\Input\ArgvInput());
```
> app/AppKernel.php

```php
<?php

use Symfony\Component\Config\Loader\LoaderInterface;
use Symfony\Component\HttpKernel\Kernel;

class AppKernel extends Kernel
{
    public function registerBundles()
    {
        return [
            new Symfony\Bundle\FrameworkBundle\FrameworkBundle(),
            new Symfony\Bundle\SecurityBundle\SecurityBundle(),
            new Doctrine\Bundle\DoctrineBundle\DoctrineBundle(),
        ];
    }

    public function registerContainerConfiguration(LoaderInterface $loader)
    {
        $loader->load(__DIR__.'/config/config.yml');
    }
}
```

> app/config/config.yml

```yml
security:
    acl:
        connection: default
    providers:
        mem:
          memory:
            users: ~
    firewalls:
        main:
            anonymous:    true

services:
      some.subscriber:
          class: SomeSubscriber
          tags:
              - { name: doctrine.event_subscriber }

parameters:
    domain: dev.enrola
    domain_port: 8280
    secret: tokenToBeGenerated

framework:
    secret: "%secret%"
    router:
        resource: "%kernel.root_dir%/config/routing.yml"
    csrf_protection: ~
    trusted_proxies: ~
    session: ~

doctrine:
  dbal:
    default_connection: default
```
> src/SomeSubscriber.php
```php
<?php

class SomeSubscriber implements Doctrine\Common\EventSubscriber
{
    public function getSubscribedEvents()
    {
        return [];
    }
}
```

First, build the cache via `bin/console cache:clear`. Then, remove (or rename) the SomeSubscriber (removing/changing the service definition is optional, code won't get as far). Then, try to clear the cache again.

**Comment by xabbuh** [comment]
The reason is that the cached container still references your old class which it will then try to instantiate during clearing the cache. The clean way to clear the cache then is to just remove its directory from the filesystem.

**Comment by ostrolucky** [comment]
>  container still references your old class which it will then try to instantiate during clearing the cache

Exactly. And it shouldn't do that when clearing the cache. `cache:clear` command ideally shouldn't use cached container at all. This issue does not occur when using `dev` environment. Second thing is, I don't think it's right that ACL causes this problem. Can't you fetch the Connection instance lazily, so it doesn't try to register listeners as soon as it is injected?

> clean way to clear the cache then is to just remove its directory from the filesystem.

When this happens, it's the only way, really.

**Comment by fabpot** [comment]
That's a known limitation of the `cache:clear` command.

**Comment by ostrolucky** [comment]
@fabpot If that's known limitation of cache:clear, why is it recommended on [ How to Deploy a Symfony Application](http://symfony.com/doc/current/deployment.html#d-clear-your-symfony-cache) page? It makes deployment very unreliable. I don't even know why this command exists if it's accepted that it's not reliable for the one specific use case it was designed for.

**Comment by sstok** [comment]
Does `cache:clear --no-warmup` and then `cache:warm` solve your problem?

**Comment by mvrhov** [comment]
@sstok it probably doesn't as the class doesn't exist anymore.
IMO symfony should suggest that the proper way of deploying is to deploy into different directory and then change symlinks. I know it's hard as you need more infrastructure and another can of worms. 
The overwrite everything method seems easy but also has its own caveats. Another one is db migrations.

**Comment by ostrolucky** [comment]
@sstok no

**Comment by nicolas-grekas** [comment]
Would be fixed by #23488
